fixed crash at mediaplayer closing, fixed tsmf event parameters-no video

AMV007 opened the pull request:
freerdp crashed at media player closing windows 2008R2
no tsmf video output at test video sample about bunny.
all this fixed.

bmiklautz commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -840,11 +854,13 @@ void tsmf_presentation_set_geometry_info(TSMF_PRESENTATION* presentation,
 
 	tmp_rects = realloc(presentation->rects, sizeof(RDP_RECT) * num_rects);
-	if (!tmp_rects)
-		return;
```
Any specific reason why you removed this check?

bmiklautz commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -375,6 +375,18 @@ static void tsmf_sample_playback_video(TSMF_SAMPLE* sample)
+		if (presentation->nr_rects > 0)
+		{
+			event.numVisibleRects = presentation->nr_rects;
+			event.visibleRects = (RECTANGLE_16*) calloc(1, event.numVisibleRects * sizeof(RECTANGLE_16));
```
result of  calloc isn't checked.

bmiklautz commented:
@AMV007 how did you test (what tsfm "backend") and how did the crash happen?

AMV007 commented:
@bmiklautz 
>>Any specific reason why you removed this check?
it not removed, just moved down, because if realloc with size 0 we must update pointer presentation->rects to new value, because memory is freed !

Added check for calloc

bmiklautz commented:
If a realloc with size 0 can happen this case should be handled separately otherwise you can't distinguish if realloc failed or if it was intended.

bmiklautz commented:
@AMV007 I would also be interested in how you did your testing.

AMV007 commented:
fixed realloc and calloc.
Tested - simple win2008r2 - downloaded big_buck_bunny_720p_h264 video from description of tsmf and tried play it with media player.

AMV007 commented:
additionally fixed tsmf - no audio

hardening commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -406,6 +426,8 @@ static void tsmf_sample_playback_video(TSMF_SAMPLE* sample)
 
 		free(event.frameData);
+
+		if(event.visibleRects!=NULL) free(event.visibleRects);
```
coding style: spaces  around NULL, and free() on its own line

hardening commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -840,8 +875,10 @@ void tsmf_presentation_set_geometry_info(TSMF_PRESENTATION* presentation,
 	tmp_rects = realloc(presentation->rects, sizeof(RDP_RECT) * num_rects);
-	if (!tmp_rects)
+
+	if (!tmp_rects&&num_rects)
```
if num_rects is 0, presentation->rects is freed so it should be set to null

hardening commented on channels/tsmf/client/tsmf_media.c:
```diff
@@ -375,6 +377,24 @@ static void tsmf_sample_playback_video(TSMF_SAMPLE* sample)
+				return;
+			}
+			memcpy(event.visibleRects, presentation->rects, presentation->nr_rects * sizeof(RDP_RECT));
+			presentation->nr_rects=0;
```
coding style: missing spaces around =

hardening commented:
except my (blocking) remarks, +1

AMV007 commented:
fixed issues from hardening

akallabeth commented:
@bmiklautz @hardening Any objections against a manual merge?

hardening commented:
@akallabeth yep no pb

akallabeth commented:
Superseeded by #3707
